Public/Src/Tools/SymbolDaemon/SymbolDaemon.cs

```diff
@@ -0,0 +1,424 @@
+                serviceEndpoint: conf.Get(ServiceEndpoint),
+                retention: TimeSpan.FromDays(conf.Get(RetentionDays)),
+                httpSendTimeout: TimeSpan.FromMilliseconds(conf.Get(HttpSendTimeoutMillis)),
+                verbose: conf.Get(Verbose),
+                enableTelemetry: conf.Get(EnableTelemetry),
+                logDir: conf.Get(LogDir));
+        }
+
+        private static Client CreateClient(string serverMoniker, IClientConfig config)
+        {
+            return serverMoniker != null
+                ? Client.Create(serverMoniker, config)
+                : null;
+        }
+
+        internal static readonly Command StartNoServiceCmd = RegisterCommand(
+            name: "start-noservice",
+            description: @"Starts a server process without a backing symbol service client (useful for testing/pinging the daemon).",
+            needsIpcClient: false,
+            clientAction: (conf, _) =>
+            {
+                var symbolConfig = new SymbolConfig(string.Empty, new Uri("file://xyz"));
+                var daemonConfig = CreateDaemonConfig(conf);
+                var vsoClientTask = TaskSourceSlim.Create<ISymbolClient>();
+                vsoClientTask.SetException(new NotSupportedException());
+                using (var daemon = new SymbolDaemon(conf.Config.Parser, daemonConfig, symbolConfig, vsoClientTask.Task))
+                {
+                    daemon.Start();
+                    daemon.Completion.GetAwaiter().GetResult();
+                    return 0;
+                }
+            });
+
+        internal static readonly Command StartCmd = RegisterCommand(
+           name: "start",
+           description: "Starts the server process.",
+           options: SymbolConfigOptions.Union(new[] { IpcServerMonikerOptional }),
+           needsIpcClient: false,
+           clientAction: (conf, _) =>
+           {
+               SetupThreadPoolAndServicePoint(s_minWorkerThreads, s_minIoThreads, s_servicePointParallelism);
+               var symbolConfig = CreateSymbolConfig(conf);
+               var daemonConf = CreateDaemonConfig(conf);
+
+               if (daemonConf.MaxConcurrentClients <= 1)
+               {
+                   conf.Logger.Error($"Must specify at least 2 '{nameof(DaemonConfig.MaxConcurrentClients)}' when running SymbolDaemon to avoid deadlock when stopping this daemon from a different client");
+                   return -1;
+               }
+
+               using (var client = CreateClient(conf.Get(IpcServerMonikerOptional), daemonConf))
+               using (var daemon = new SymbolDaemon(
+                   parser: conf.Config.Parser,
+                   daemonConfig: daemonConf,
+                   symbolConfig: symbolConfig,
+                   symbolServiceClientTask: null,
+                   bxlClient: client))
+               {
+                   daemon.Start();
+                   daemon.Completion.GetAwaiter().GetResult();

```

Comment:
>.GetAwaiter().GetResult( [](start = 36, length = 24)

Doesn't the clietn action support async lambda's like below with CreateSymbolRequestCmd?

Reply:
This comment is a little bit special - it will be used by bxl to create a service pip. It will be called from Program.Main and we must wait for it to complete here.

---
In reply to: [313489440](https://github.com/microsoft/BuildXL/pull/727#discussion_r313489440) [](ancestors = 313489440)